swirl/src/runner.rs

```diff
@@ -85,42 +86,91 @@ impl<Env, ConnectionPool> Runner<Env, ConnectionPool> {
 impl<Env, ConnectionPool> Runner<Env, ConnectionPool>
 where
     Env: RefUnwindSafe + Send + Sync + 'static,
-    ConnectionPool: DieselPoolOwned + 'static,
+    ConnectionPool: DieselPool + 'static
 {
-    pub fn run_all_pending_jobs(&self) -> Result<(), PerformError> {
-        if let Some(conn) = self.try_connection() {
-            let available_job_count = storage::available_job_count(&conn)?;
-            for _ in 0..available_job_count {
-                self.run_single_job()
+    /// Runs all pending jobs in the queue
+    ///
+    /// This function will return once all jobs in the queue have begun running,
+    /// but does not wait for them to complete. When this function returns, at
+    /// least one thread will have tried to acquire a new job, and found there
+    /// were none in the queue.
+    pub fn run_all_pending_jobs(&self) -> Result<(), FetchError<ConnectionPool>> {
+        use std::cmp::max;
+
+        let max_threads = self.thread_pool.max_count();
+        let (sender, receiver) = channel::new(max_threads);
+        let mut pending_messages = 0;
+        loop {
+            let available_threads = max_threads - self.thread_pool.active_count();
+
+            let jobs_to_queue = if pending_messages == 0 {
+                // If we have no queued jobs talking to us, and there are no
+                // available threads, we still need to queue at least one job
+                // or we'll never receive a message
+                max(available_threads, 1)
+            } else {
+                available_threads
+            };
+
+            for _ in 0..jobs_to_queue {
+                self.run_single_job(sender.clone());
+            }
+
+            pending_messages += 1;
+            match receiver.recv() {
+                Ok(Event::Working) => pending_messages -= 1,
```

Comment:
I'm not sure I understand how `pending_messages` works here.  It seems like the only way for the loop to continue (all other paths `return`) is for `pending_messages` to always be decremented back to 0.

I *think* I'm expecting something more like `pending_messages += jobs_to_queue`, and then looping over the match until `pending_messages == 0`.

Reply:
See the last two paragraphs of the commit message. We return as soon as any worker reports no jobs left because this helps better keep our work threads saturated, and the only times we would receive that message with `pending_messages > 0` are:

- we're about to get a lot more messages saying no jobs are left
- an extremely unlikely error occurred that we'll just get again the next time this function is called

Reply:
`pending_messages` is basically just so we know whether we need to enqueue a single job even if there are no available threads. Otherwise we could get 8 working messages and loop again before any job completes (so no threads are idle). In that case we need to make sure we queue another job to run or we will never get a message back